subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/InstrumentingBuildService.java

```diff
@@ -0,0 +1,90 @@
+import org.gradle.internal.classpath.types.InstrumentingTypeRegistry;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class InstrumentingBuildService implements BuildService<InstrumentingBuildService.InstrumentBuildServiceParameters> {
+
+    public interface InstrumentBuildServiceParameters extends BuildServiceParameters {
+        ConfigurableFileCollection getClassHierarchy();
+    }
+
+    private final AtomicReference<InstrumentingTypeRegistry> instrumentingTypeRegistry = new AtomicReference<>();
+
+    public InstrumentingTypeRegistry getInstrumentingTypeRegistry(GradleCoreInstrumentingTypeRegistry gradleCoreInstrumentingTypeRegistry) {
+        if (gradleCoreInstrumentingTypeRegistry.isEmpty()) {
+            return InstrumentingTypeRegistry.empty();
+        }
+        if (instrumentingTypeRegistry.get() != null) {
+            return instrumentingTypeRegistry.get();
+        }
+        return newInstrumentingTypeRegistry(gradleCoreInstrumentingTypeRegistry);
+    }
+
+    private synchronized InstrumentingTypeRegistry newInstrumentingTypeRegistry(GradleCoreInstrumentingTypeRegistry gradleCoreInstrumentingTypeRegistry) {
+        if (instrumentingTypeRegistry.get() != null) {
+            return instrumentingTypeRegistry.get();
+        }
+        Map<String, Set<String>> classHierarchy = readClassHierarchy();
+        ExternalPluginsInstrumentingTypeRegistry registry = new ExternalPluginsInstrumentingTypeRegistry(classHierarchy, gradleCoreInstrumentingTypeRegistry);
+        instrumentingTypeRegistry.set(registry);
+        return registry;
+    }
+
+    private Map<String, Set<String>> readClassHierarchy() {
+        Set<File> files = getParameters().getClassHierarchy().getFiles();
+        Map<String, Set<String>> classHierarchy = new HashMap<>();
+        files.forEach(file -> {
+            Properties properties = new Properties();
+            try (InputStream inputStream = Files.newInputStream(file.toPath())) {
+                properties.load(inputStream);
+                properties.forEach((key, value) -> {
+                    Set<String> keyTypes = classHierarchy.computeIfAbsent((String) key, __ -> new HashSet<>());
+                    Collections.addAll(keyTypes, ((String) value).split(","));
+                });
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+        return classHierarchy;
+    }
+
+    public void clear() {
+        // Remove the reference to the registry, so that it can be garbage collected,
+        // since build service instance is not deregistered
+        instrumentingTypeRegistry.set(null);
```

Comment:
🤔 This looks real fishy. Can't we have build services be closed properly?

Reply:
See a comment https://github.com/gradle/gradle/pull/27698#discussion_r1482155946.

Reasoning is that we don't need data after artifact transform is done, so we can clear everything immediately after artifact transform is done. And we don't wait to the end of the build.

Reply:
This now seems to be gone.